source/MaterialXGenShader/Util.cpp

```diff
@@ -508,4 +509,105 @@ void mapValueToColor(const ValuePtr value, Color4& color)
     }
 }
 
+bool requiresImplementation(const NodeDefPtr nodeDef)
+{
+    if (!nodeDef)
+    {
+        return false;
+    }
+    static std::string TYPE_NONE("none");
+    const std::string typeAttribute = nodeDef->getAttribute(TypedElement::TYPE_ATTRIBUTE);
+    return !typeAttribute.empty() && typeAttribute != TYPE_NONE;
+}
+
+bool elementRequiresShading(const TypedElementPtr element)
+{
+    std::string elementType(element->getType());
+    const std::set<std::string> colorClosures =
+    {
+        "surfaceshader", "volumeshader", "lightshader",
+        "BSDF", "EDF", "VDF"
+    };
+    return (element->isA<ShaderRef>() ||
+            colorClosures.count(elementType) > 0);
+}
+
+void findRenderableElements(const DocumentPtr& doc, std::vector<TypedElementPtr>& elements)
```

Comment:
This is the main extraction. materialXView has a rough copy of this logic currently so want to make it reusable so it's consistent.